server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -276,38 +281,7 @@ private synchronized boolean assertNoSeqNumberConflict(long seqNo, BytesReferenc
                 Translog.Operation prvOp = Translog.readOperation(
                     new BufferedChecksumStreamInput(previous.v1().streamInput(), "assertion")
                 );
-                // TODO: We haven't had timestamp for Index operations in Lucene yet, we need to loosen this check without timestamp.
-                final boolean sameOp;
-                if (newOp instanceof final Translog.Index o2 && prvOp instanceof final Translog.Index o1) {
-                    sameOp = Objects.equals(o1.id(), o2.id())
-                        && Objects.equals(o1.source(), o2.source())
-                        && Objects.equals(o1.routing(), o2.routing())
-                        && o1.primaryTerm() == o2.primaryTerm()
-                        && o1.seqNo() == o2.seqNo()
-                        && o1.version() == o2.version();
-                } else if (newOp instanceof final Translog.Delete o1 && prvOp instanceof final Translog.Delete o2) {
-                    sameOp = Objects.equals(o1.id(), o2.id())
-                        && o1.primaryTerm() == o2.primaryTerm()
-                        && o1.seqNo() == o2.seqNo()
-                        && o1.version() == o2.version();
-                } else {
-                    sameOp = false;
-                }
-                if (sameOp == false) {
-                    throw new AssertionError(
-                        "seqNo ["
-                            + seqNo
-                            + "] was processed twice in generation ["
-                            + generation
-                            + "], with different data. "
-                            + "prvOp ["
-                            + prvOp
-                            + "], newOp ["
-                            + newOp
-                            + "]",
-                        previous.v2()
-                    );
-                }
+                assert operationAsserter.assertSameOperations(seqNo, generation, newOp, prvOp, previous.v2());
```

Comment:
I wonder if we should change `assertSameOperations` to just return a boolean and then add the message here instead, avoiding to pass a few params to `assertSameOperations`? Something like:

```suggestion
                assert operationAsserter.sameOperations(newOp, prvOp) : new AssertionError(
                        "seqNo ["
                            + seqNo
                            + "] was processed twice in generation ["
                            + generation
                            + "], with different data. "
                            + "prvOp ["
                            + prvOp
                            + "], newOp ["
                            + newOp
                            + "]",
                        previous.v2()
                    );
```

Reply:
++ see https://github.com/elastic/elasticsearch/pull/119330/commits/cc259d21ce9d53cae70168bb92803e54529b8a6c